### PR TITLE
remove dependency on aligned_storage and aligned_union

### PR DESCRIPTION
**Pull request by arvidn**
to support older versions of GCC

### CODE REVIEW

**Comment by aldenml** [comment]
the ubuntu toolchain PPA supports gcc 5 for very old dist versions, don't you think it's better to simply drop support for gcc <= 5.1? gcc 4 is no longer maintained

**Comment by arvidn** [comment]
I don't know what PPA means.. just that sometimes systems have old toolchains. cas-- points out that gcc 4.8 is supported in Ubuntu Trusty until 2019. I take it that's the LTS version perhaps.

**Comment by aldenml** [comment]
sorry I assumed some familiarity...no need of complicating this if you already made it

**Comment by arvidn** [comment]
it seems msvc makes this a bit trickier

File: src/cpuid.cpp [hunk 1]
```diff
@@ -46,7 +46,7 @@ POSSIBILITY OF SUCH DAMAGE.
 #include <cstring> // for std::memset
 #endif
 
-#if TORRENT_HAS_ARM
+#if TORRENT_HAS_ARM && TORRENT_HAS_ARM_NEON
```
Comment (arvidn) [inline src/cpuid.cpp]: @aldenml does this seem reasonable? as far as I can tell, this function is only used if ``TORRENT_HAS_ARM_NEON``, and in my cross compiler for ``arm-linux-gnueabihf`` does not have this header

Comment (aldenml) [inline src/cpuid.cpp]: It is used for `TORRENT_HAS_ARM_CRC32` too, I think you can add it to this condition and solve the issue. What toolchain/sysroot are you using?

Comment (arvidn) [inline src/cpuid.cpp]: linaro for raspberry pi. arm-linux-gnueabihf-raspbian
Additionally, this is a linux header, so presumably it should only be used on ``__linux__`` as well, right?

Comment (aldenml) [inline src/cpuid.cpp]: true.

btw, linaro is the same you have in travis and raspberry pi is a linux based system :)

Comment (arvidn) [inline src/cpuid.cpp]: this header wasn't introduced until glibc 2.16 as far as I can tell, so I added that condition. It doesn't seem linux specific, just glibc specific

File: src/cpuid.cpp [hunk 2]
```diff
@@ -46,7 +46,14 @@ POSSIBILITY OF SUCH DAMAGE.
 #include <cstring> // for std::memset
 #endif
 
-#if TORRENT_HAS_ARM && defined __linux__ && (TORRENT_HAS_ARM_NEON || TORRENT_HAS_ARM_CRC32)
+#if defined __GLIBC__ && (__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 16))
```
Comment (aldenml) [inline src/cpuid.cpp]: do you mind adding here `|| TORRENT_ANDROID`?

my understanding is that it was added in `glibc` but it's _not_ `glibc` specific, in fact, that's the documented way in android to get the CPU features.

Comment (arvidn) [inline src/cpuid.cpp]: isn't android using glibc though?

Comment (aldenml) [inline src/cpuid.cpp]: no, it's `bionic`

**Pull request merged**